OPHD/StructureManager.cpp

```diff
@@ -244,21 +244,31 @@ void StructureManager::addStructure(Structure* structure, Tile* tile)
  */
 void StructureManager::removeStructure(Structure* structure)
 {
+	// Destroy all components belonging to the structure.
+	SKey s = SKey(structure);
+	for (auto& [componentTypeID, table] : mComponents)
+	{
+		auto cit = table.find(s);
+		if (cit != table.end())
+		{
+			delete cit->second;
```

Comment:
Naked `new`/`delete` should probably be avoided in C++17 code and onwards. In the general case, they can be avoided by using [`std::unique_ptr`](https://en.cppreference.com/w/cpp/memory/unique_ptr) instead. Though in more specific cases, which I believe applies here, you could update the collection class to instead store the object itself rather than a pointer to it. In that case the collection class is responsible for allocating and freeing associated object memory. I would recommend investigating the later approach of using the collection class to manage the object memory.

----

I see no corresponding `new` anywhere in this code, which makes me wonder if there's a memory management issue.

Reply:
The memory management here follows the exact same pattern as what was being done for Structure: StructureCatalog is expected to create the components, and their ownership are transferred to StructureManager as they are registered. The comment for StructureManager::attachComponent explains this.

Unfortunately you can't make the container store the object unless you have a separate member for each type, which I would only do if I were to include an .inc file with macros listing each component so I could generate the member table without increasing maintenance burden. I assume you would dislike such a solution. Explicitly listing one member per type is also less flexible in loading additional component types from extension DLLs or whatnot.

Reply:
rmtoth is correct -- memory is not allocated in `StructureManager`, it's allocated elsewhere (typically from `StructureCatalog`). It would make sense to use the delete operator here though it could be useful to use smart pointers instead. That would probably take a refactor of the `StructureCatalog` and anywhere else structures are instantiated and probably worth being a separate change set.

Reply:
Ahh, I'm not totally familiar with some aspects of this part of the code.

I agree this should be addressed as a separate change set. I believe #480 was opened to address design considerations such as this.

Though I suspect the ECS system could potentially obsolete some of the code involved here. It would be possible to have `std::vector` of the various `Component` subtypes, rather than a single `std::vector` with pointers to the base type. Not sure if that's how it would be implemented in the end though.

Reply:
There are several reasons why you'd want to iterate over all the component types of a structure and use virtual functions from the StructureComponent base type:
* When destroying a structure, you need to destroy all its components.
* When populating the inspector view of a structure, you want to give every component a chance to add information.
* When saving the game, you need to serialize all the components of a structure.
* When deserializing the game, you need to insert all components of a structure.
* When a structure changes state, you may want to notify all its components.

I'm sure there will be many more cases. I don't think it's good design if you have to manually add a newly implemented component type to each of these locations in the codebase.

That said, changing the backing storage to per-component-type vectors (which could potentially live in the component classes themselves) does not preclude a way of iterating over all of them. But I don't see why you'd go to such lengths just to avoid using a cast and a virtual destructor.

Reply:
I'm wondering if that could be handled by iterating over a `tuple`. Maybe a tuple of collections of the various types?

Anyway, I would agree you don't want to be manually updating all those places when adding new component types.